pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1171,34 +1186,50 @@ func (c *Reconciler) updatePipelineRunStatusFromInformer(ctx context.Context, pr
 		logger.Errorf("could not list TaskRuns %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromTaskRuns(logger, pr, taskRuns)
+	updatePipelineRunStatusFromTaskRuns(ctx, logger, pr, taskRuns)
 
 	runs, err := c.runLister.Runs(pr.Namespace).List(k8slabels.SelectorFromSet(pipelineRunLabels))
 	if err != nil {
 		logger.Errorf("could not list Runs %#v", err)
 		return err
 	}
-	updatePipelineRunStatusFromRuns(logger, pr, runs)
+	updatePipelineRunStatusFromRuns(ctx, logger, pr, runs)
 
 	return nil
 }
 
-func updatePipelineRunStatusFromTaskRuns(logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
+func updatePipelineRunStatusFromTaskRuns(ctx context.Context, logger *zap.SugaredLogger, pr *v1beta1.PipelineRun, trs []*v1beta1.TaskRun) {
 	// If no TaskRun was found, nothing to be done. We never remove taskruns from the status
 	if trs == nil || len(trs) == 0 {
 		return
 	}
+
+	cfg := config.FromContextOrDefaults(ctx)
+	fullEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.FullEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+	minimalEmbedded := cfg.FeatureFlags.EmbeddedStatus == config.MinimalEmbeddedStatus || cfg.FeatureFlags.EmbeddedStatus == config.BothEmbeddedStatus
+
 	// Store a list of Condition TaskRuns for each PipelineTask (by name)
 	conditionTaskRuns := make(map[string][]*v1beta1.TaskRun)
 	// Map PipelineTask names to TaskRun names that were already in the status
 	taskRunByPipelineTask := make(map[string]string)
-	if pr.Status.TaskRuns != nil {
-		for taskRunName, pipelineRunTaskRunStatus := range pr.Status.TaskRuns {
-			taskRunByPipelineTask[pipelineRunTaskRunStatus.PipelineTaskName] = taskRunName
+	// Map PipelineTask names to TaskRun child references that were already in the status
+	childRefByPipelineTask := make(map[string]*v1beta1.ChildStatusReference)
+
+	if fullEmbedded {
+		if pr.Status.TaskRuns != nil {
+			for taskRunName, pipelineRunTaskRunStatus := range pr.Status.TaskRuns {
+				taskRunByPipelineTask[pipelineRunTaskRunStatus.PipelineTaskName] = taskRunName
+			}
+		} else {
+			pr.Status.TaskRuns = make(map[string]*v1beta1.PipelineRunTaskRunStatus)
```

Comment:
I think it is - if I remember correctly (the logic here is a bit of a mess and I can't remember all the details off the top of my head!), `pr.Status.TaskRuns` does start out nil, but everything below here expects it to be a created map, so in the cases where `embedded-status` is either `full` or `both`, we need to initialize `pr.Status.TaskRuns` so it can be properly interacted with later.

Reply:
I don't think we should be treating nil maps differently from empty maps-- if that's the case that logic should be changed instead of initializing the variable to an empty map.

Reply:
I'll need to dig to make sure changing that makes sense.

Reply:
Ok, yeah, I think initializing the map makes sense in this situation. If we've gotten this far (i.e., `len(trs) > 0` or we would have returned out right away at the top of the function, and `fullEmbedded` is true, meaning `embedded-status` is either `full` or `minimal`, and nothing has initialized `pr.Status.TaskRuns` yet, because it's `nil`), we know that there are `TaskRun`s to process, we know that we will be populating them into `pr.Status.TaskRuns`. So the map will have to be initialized at some point for sure, and it's cleaner to do it here than in the loop, I feel.